### DISCUSSION TITLE
Cloudinary Storage Driver: Configurable Asset Transformations on Upload

### DISCUSSION THREAD

**Discussion by sethkaufee** [directus/directus#19311] [Feature Requests]
### Summary

The Cloudinary API has the ability to create transformations at upload time: [Link to Docs](https://cloudinary.com/documentation/transformations_on_upload) 

It would be very beneficial to expose this Cloudinary API via the Directus storage configuration by passing an "upload preset" parameter: [Link to Docs](https://cloudinary.com/documentation/image_upload_api_reference#upload)

This parameter references a preset that is created within the Cloudinary Admin.

### Basic Example

This would be an example configuration parameter that would include the "hls-encode" preset to all uploaded files, configured on Cloudinary to generate HLS encodings to the uploaded video assets:

```env
STORAGE_LOCATIONS="cloudinary"
STORAGE_CLOUDINARY_DRIVER: "cloudinary"
STORAGE_CLOUDINARY_ACCESS_MODE="public"
STORAGE_CLOUDINARY_UPLOAD_PRESETS=hls-encode
```

Or the developer can pass a glob pattern to only include the preset to certain uploaded files:
```env
STORAGE_LOCATIONS="cloudinary"
STORAGE_CLOUDINARY_DRIVER: "cloudinary"
STORAGE_CLOUDINARY_ACCESS_MODE="public"
STORAGE_CLOUDINARY_UPLOAD_PRESETS='json:{"hls-encode": "*.mp4"}'
```

### Motivation

Cloudinary offers many methods to transform and generate additional assets that Directus does not offer or interface with, especially at upload time. The ability to access more of Cloudinary's api and capabilities in would be extremely useful. Including this preset value on uploads (if opted-in), would add much more value/features to the Cloudinary driver while also making it easy to implement and maintain over time.

### Detailed Design

Normally if the Cloudinary driver is enabled, a single uploaded file to Cloudinary will generate an asset. 

This feature will allow users to add an additional configuration, which will add a parameter to the request:
```env
STORAGE_LOCATIONS="cloudinary"
STORAGE_CLOUDINARY_DRIVER: "cloudinary"
STORAGE_CLOUDINARY_ACCESS_MODE="public"
STORAGE_CLOUDINARY_UPLOAD_PRESETS=SOME-PRESET # new parameter
```

In the storage driver code:
```typescript
class DriverCloudinary implements Driver {
    constructor(config: DriverCloudinaryConfig) {
         // handle new configuration param on initialization
    }
    async write(filepath: string, content: Readable) {
         const uploadParameters = {
             public_id: "....",
             resource_type: "...",
            // ...
         }
         
         // check if upload preset exists and matches with filename/path if configuration is a glob pattern
         if (this.getUploadPreset(filename)) {
             uploadParameters.upload_preset = this.getUploadPreset(filename);
         }

         // Calls upload api
    }
}
```

### Must Have:
- No breaking changes to existing Cloudinary integrations
- Ability to attach an upload preset for uploads, which is configurable in the Cloudinary storage driver configuration.

### Should Have:
- 

### Could Have:
-

### Won't Have:
- Ability to link any Cloudinary generated assets to Directus. This is intended to be used at the user's discretion. 

### Drawbacks

- Some upload preset's could be incompatible with Directus or break the asset linking. 

### Alternatives

- Include more fine grain transformations in the Directus configuration, but the major drawback is that is could change over time and/or be difficult to maintain the driver. 

### Adoption Strategy

This is not a breaking change. Directus users would be able to add this new configuration at their discretion, knowing that their preset could be incompatible with Directus, i.e. using a prefix that changes the file name, folder, or asset.

**Comment by sethkaufee** [comment]
[Link to working POC](https://github.com/sethkaufee/directus.git#discussions/19311)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
